### PR TITLE
Using `ctr import` instead of pulling images

### PR DESCRIPTION
**Pull request by turan18**
This change allows all images that are needed within our integration tests to be pulled during integration setup rather than test runtime. Since our integration tests run in isolated containers we could docker `COPY` the tarballs for each image (obtained via `ctr export`) for all the test images into our test container, whereby each test container could then `ctr image import` them without having to pull them.

**The main intention behind this PR was to speed up integration test times. Looking at both GitHub runners integ test time as well as local integ test times it appears that importing, surprisingly, does not seem to provide a faster test runtime in most cases. I'm converting this PR to a draft for now, while I explore this issue further**

*Issue #, if available:*

Fixes: #385

*Description of changes:*

*Testing performed:*

`make test && make integration`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -24,7 +24,8 @@ ARG RUNC_VERSION
 ARG NERDCTL_VERSION
 COPY . $GOPATH/src/github.com/awslabs/soci-snapshotter
 ENV GOPROXY direct
-RUN apt-get update -y && apt-get install -y libbtrfs-dev libseccomp-dev libz-dev gcc fuse pigz
+RUN mkdir testImages/ && apt-get update -y && apt-get install -y libbtrfs-dev libseccomp-dev libz-dev gcc fuse pigz
+COPY integration/_images/ testImages/
```
Comment (sparr) [inline Dockerfile:28]: mkdir should be unnecessary. COPY should create destination directories.

File: Makefile [hunk 2]
```diff
@@ -76,7 +76,7 @@ test:
 integration: build
 	@echo "$@"
 	@echo "SOCI_SNAPSHOTTER_PROJECT_ROOT=$(SOCI_SNAPSHOTTER_PROJECT_ROOT)"
-	@GO111MODULE=$(GO111MODULE_VALUE) SOCI_SNAPSHOTTER_PROJECT_ROOT=$(SOCI_SNAPSHOTTER_PROJECT_ROOT) ENABLE_INTEGRATION_TEST=true go test $(GO_TEST_FLAGS) -v -timeout=0 ./integration
+	@GO111MODULE=$(GO111MODULE_VALUE) sudo SOCI_SNAPSHOTTER_PROJECT_ROOT=$(SOCI_SNAPSHOTTER_PROJECT_ROOT) ENABLE_INTEGRATION_TEST=true go test $(GO_TEST_FLAGS) -v -timeout=0 ./integration
```
Comment (sparr) [inline Makefile:79]: Why is this necessary?

File: integration/testutil.go [hunk 3]
```diff
@@ -320,9 +332,12 @@ func encodeImageInfoNerdctl(ii ...imageInfo) [][]string {
...
 func copyImage(sh *shell.Shell, src, dst imageInfo) {
 	opts := encodeImageInfoNerdctl(src, dst)
-	sh.
-		X(append([]string{"nerdctl", "pull", "-q", "--platform", platforms.Format(src.platform)}, opts[0]...)...).
-		X("ctr", "i", "tag", src.ref, dst.ref).
+	if src.mirror {
+		sh.X(append([]string{"nerdctl", "pull", "-q", "--platform", platforms.Format(src.platform)}, opts[0]...)...)
+	} else {
+		sh.X(append([]string{"ctr", "i", "import", "--platform", platforms.Format(src.platform)}, fmt.Sprintf("testImages/%s.tar", src.imgName))...)
+	}
```
Comment (sparr) [inline integration/testutil.go:339]: this pattern appears 3 places, can we move it to a helper?

File: integration/testutil.go [hunk 4]
```diff
@@ -703,7 +726,66 @@ func setup() ([]func() error, error) {
...
 
-	return compose.Build(composeYaml, cOpts...)
+	ctx := namespaces.WithNamespace(context.Background(), "soci-testing")
+	client, err := containerd.New(containerdAddress)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to containerd: %w", err)
+	}
+	defer client.Close()
+
+	testImages := []string{alpineImage, ubuntuImage, drupalImage, nginxImage, rabbitmqImage, rabbitmqShaImage, rethinkdbShaImage}
```
Comment (sparr) [inline integration/testutil.go:736]: Future additions to the image constants might not get added here. At minimum, I suggest adding a comment near https://github.com/awslabs/soci-snapshotter/pull/428/files#diff-a1077885a297cb6bcfbde19b03d7c04d94209ca38a3210644c969726aff4d985R82 pointing here. More preferably, move the creation of this slice there.

File: integration/testutil.go [hunk 5]
```diff
@@ -77,13 +81,14 @@ const (
...
-	ubuntuImage   = "ubuntu:23.04"
-	drupalImage   = "drupal:10.0.2"
-	rabbitmqImage = "rabbitmq:3.11.7"
+	alpineImage       = "alpine:3.17.1"
+	nginxImage        = "nginx:1.23.3"
+	ubuntuImage       = "ubuntu:23.04"
+	drupalImage       = "drupal:10.0.2"
+	rabbitmqImage     = "rabbitmq:3.11.7"
+	rabbitmqShaImage  = "rabbitmq@sha256:603be6b7fd5f1d8c6eab8e7a234ed30d664b9356ec1b87833f3a46bb6725458e"
+	rethinkdbShaImage = "rethinkdb@sha256:4452aadba3e99771ff3559735dab16279c5a352359d79f38737c6fdca941c6e5"
```
Comment (sbuckfelder) [inline integration/testutil.go:90]: Why the additional two images?  why different format?

File: integration/testutil.go [hunk 6]
```diff
@@ -703,7 +726,66 @@ func setup() ([]func() error, error) {
...
+			fmt.Println("pulling", ref, "...")
+			i, err = client.Fetch(ctx, ref, remoteOpts...)
+			if err != nil {
+				return nil, fmt.Errorf("failed to pull image %s: %w", ref, err)
+			}
+		} else {
+			fmt.Printf("%s already exists. skipping pulling...\n", i.Name)
+		}
+		fmt.Println("exporting...")
+		tarballPath := filepath.Join(testImagePath, fmt.Sprintf("%s.tar", img[:strings.IndexByte(img, ':')]))
```
Comment (sbuckfelder) [inline integration/testutil.go:769]: isn't this in the ref.imageName ?

**Comment by sbuckfelder** [comment]
Did this work?  What kind of speed pick up did you see on the integration tests?

File: integration/testutil.go [hunk 7]
```diff
@@ -290,14 +295,21 @@ func withPlatform(p spec.Platform) imageOpt {
...
+	mirror    bool
 	creds     string
 	plainHTTP bool
 	platform  spec.Platform
 }
 
 func dockerhub(name string, opts ...imageOpt) imageInfo {
-	i := imageInfo{dockerLibrary + name, "", false, platforms.DefaultSpec()}
+	i := imageInfo{
+		imgName: name[:strings.IndexByte(name, ':')],
```
Comment (djdongjin) [inline integration/testutil.go:308]: this doesn't work if the image name contains a registry port (e.g. `localhost:5000/rabbitmq:latest` in our get start doc).

File: integration/testutil.go (hunk 3 above)
Comment (djdongjin) [inline integration/testutil.go:339]: can we have an estimate how many times `"nerdctl", "pull", "-q"` and `"ctr", "i", "import"` are called respectively? (e.g., you can run integration test and print `nerdctl-pull-count` (or `ctr-i-import-count`) whenever they're called and check the output.

File: integration/testutil.go [hunk 8]
```diff
@@ -721,8 +803,12 @@ func teardown(cleanups []func() error) error {
 // middleSizeLayerInfo finds a layer not the smallest or largest (if possible), returns index, size, and layer count
 // It requires containerd to be running
 func middleSizeLayerInfo(t *testing.T, sh *shell.Shell, image imageInfo) (int, int64, int) {
-	sh.O("nerdctl", "pull", "-q", "--platform", platforms.Format(image.platform), image.ref)
+	if image.mirror {
+		sh.O("nerdctl", "pull", "-q", "--platform", platforms.Format(image.platform), image.ref)
+	} else {
+		sh.O(append([]string{"ctr", "i", "import", "--platform", platforms.Format(image.platform)}, fmt.Sprintf("testImages/%s.tar", image.imgName))...)
```
Comment (djdongjin) [inline integration/testutil.go:809]: why does only these 3 places have this `if else` logic (e.g. why not other `nerdctl pull` usage)

File: integration/testutil.go (hunk 5 above)
Comment (sparr) [reply integration/testutil.go:90]: They are used in that format in existing tests.

**Review by djdongjin — COMMENTED** [review]
Personally I feel this complicates our integration test and adds some hidden steps (e.g. 1/ how to add a new image; 2/ when we should use `nerdctl pull` v.s. `ctr i import`; 3/ how to add a new platform of an existing image (e.g. in the testImages slice`)

I'd like to see more concrete improvement from the change (e.g. integration test runtime reduction).

**Comment by sparr** [comment]
> Did this work? What kind of speed pick up did you see on the integration tests?

I ran the test suite as well, but the change in time was below the noise floor for the few iterations I ran.

However, this isn't just about speed. This is also about reducing network usage, both on our end and to avoid throttling on the remote end (which was more likely with docker than ECR)

File: Makefile (hunk 2 above)
Comment (turan18) [reply Makefile:79]: Since I'm connecting to the containerd-grpc server via the client to pull images on the host.

File: integration/testutil.go (hunk 3 above)
Comment (turan18) [reply integration/testutil.go:339]: Last time I checked nerdctl was called roughly ~50 times or so. I'll run it again and check

File: integration/testutil.go (hunk 7 above)
Comment (sparr) [reply integration/testutil.go:308]: A helper to break apart the image ref into components would be valuable. I also suspect containerd might already have one.

File: integration/testutil.go (hunk 5 above)
Comment (turan18) [reply integration/testutil.go:90]: They are referenced in our pull test and create test. https://github.com/awslabs/soci-snapshotter/blob/main/integration/create_test.go#L59 and https://github.com/awslabs/soci-snapshotter/blob/main/integration/push_test.go#L90

File: integration/testutil.go (hunk 7 above)
Comment (turan18) [reply integration/testutil.go:308]: Good point , although calls to `dockerhub` assume the image to be residing in ECR. Calls to `mirror()` would break though if our local registry happens to contain ':'

File: integration/testutil.go (hunk 3 above)
Comment (turan18) [reply integration/testutil.go:339]: Just ran the integration suite locally here and noted the counts along with the time it took for each operation (pull + import). Here are the results:

pull count: 45
import count: 83
avg pull time: 776.2 ms
avg import time: [4305.6](tel:43058674698) ms

Comment (turan18) [reply integration/testutil.go:339]: Looks like imports generally take longer than most pulls , which is very interesting to say the least.

File: integration/testutil.go (hunk 7 above)
Comment (turan18) [reply integration/testutil.go:315]: Actually looking back, this shouldnt break at all since we are splitting the image name+tag IE (rabbitmq:1.x or rabbitmq@sha256) and not the ref IE (localhost:5000/rabbitmq:latest)

File: integration/testutil.go (hunk 3 above)
Comment (sparr) [reply integration/testutil.go]: Following up on our meeting regarding this PR. I don't think those comparison statistics convey what they seem to. The scenarios where pull and import are happening could be wildly different, e.g. one might be all small images and the other all large.

Suggestion: Disable this PR and just include the counts and timing collection. Run the integration test suite.
Then compare those results to the results you have here.
